### PR TITLE
8308592: Framework for CA interoperability testing

### PR DESCRIPTION
**Pull request by rhalade**
The new approach uses test URLs directly to verify interoperability with CA infrastructure. This would help us avoid having regular test fixes to update test artifacts as long as CAs keep test domains up to date.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8308592](https://bugs.openjdk.org/browse/JDK-8308592): Framework for CA interoperability testing (**Enhancement** - P3)

### Reviewers
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14252/head:pull/14252` \
`$ git checkout pull/14252`

Update a local copy of the PR: \
`$ git checkout pull/14252` \
`$ git pull https://git.openjdk.org/jdk.git pull/14252/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14252`

View PR using the GUI difftool: \
`$ git pr show -t 14252`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14252.diff">https://git.openjdk.org/jdk/pull/14252.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14252#issuecomment-1570963550)

### CODE REVIEW

**Comment by rhalade** [comment]
Only Actalis and BuyPass CA tests are updated with latest design. Will update others soon.

**Comment by rhalade** [comment]
> 

There are few minor differences but most of them will all look the same. I kept them separate so we can treat each CA differently and run / problem list at granular level. I will try to reduce the duplicate code further.

**Comment by seanjmullan** [comment]
I suggest changing the title and description of the bug as the fix is more significant than updating a few expired certificates.

File: test/jdk/security/infra/java/security/cert/CertPathValidator/certification/ValidatePathWithURL.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+            cacerts.load(fis, null);
+        }
+
+        rootCertificate = (X509Certificate) cacerts.getCertificate(caAlias);
+        rootPrincipal = rootCertificate.getSubjectX500Principal();
+    }
+
+    public static void enableOCSPOnly() {
+        System.setProperty("com.sun.security.enableCRLDP", "false");
+        Security.setProperty("ocsp.enable", "true");
```
Comment (seanjmullan) [inline test/jdk/security/infra/java/security/cert/CertPathValidator/certification/ValidatePathWithURL.java:71]: Technically I think you could avoid setting these properties (and running in othervm mode) and instead init your `SSLContext` with a `TrustManager` and a `CertPathTrustManagerParameters` containing a `PKIXParameters` with a `PKIXRevocationChecker` configured with the desired revocation settings. However, AFAIK, you can't pass in an `SSLContext` via the `HttpsURLConnection` API, so you would need to use `HttpsClient` instead (which is only available in Java 11 and up). However, I think this is something to consider as a future enhancement, as you have more control over the various TLS settings.

File: test/jdk/security/infra/java/security/cert/CertPathValidator/certification/ValidatePathWithURL.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        HttpsURLConnection httpsURLConnection = null;
+        try {
+            URL url = new URL(testURL);
+            httpsURLConnection = (HttpsURLConnection) url.openConnection();
+            httpsURLConnection.setInstanceFollowRedirects(false);
+            httpsURLConnection.connect();
+
+            // certain that test certificate anchors to trusted CA for VALID certificate
+            // if the connection is successful
+            Certificate[] chain = httpsURLConnection.getServerCertificates();
```
Comment (seanjmullan) [inline test/jdk/security/infra/java/security/cert/CertPathValidator/certification/ValidatePathWithURL.java:120]: Mostly FYI - but another way to get the chain is to call the `getCertPath()` method of the thrown `CertPathValidatorException`. This doesn't include the root though.

File: test/jdk/security/infra/java/security/cert/CertPathValidator/certification/ValidatePathWithURL.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            httpsURLConnection.disconnect();
+            X509Certificate interCert = null;
+
+            // fail if there is no intermediate CA or self-signed
+            if (chain.length < 2) {
+                throw new RuntimeException("Cert chain too short " + chain.length);
+            } else {
+                System.out.println("Finding intermediate certificate issued by CA");
+                for (Certificate cert: chain){
+                    if(cert instanceof X509Certificate certificate) {
```
Comment (seanjmullan) [inline test/jdk/security/infra/java/security/cert/CertPathValidator/certification/ValidatePathWithURL.java]: Nit: space after `if`.

File: test/jdk/security/infra/java/security/cert/CertPathValidator/certification/ValidatePathWithURL.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            }
+
+            throw new RuntimeException("Unhandled exception", e);
+        } catch (SSLException e) {
+            // thrown if root CA is not included in cacerts
+            throw new RuntimeException(e);
+        } catch (IOException e) {
+            throw new SkippedException("Network setup issue, skip this test", e);
+        } finally {
+            if(httpsURLConnection != null) {
```
Comment (seanjmullan) [inline test/jdk/security/infra/java/security/cert/CertPathValidator/certification/ValidatePathWithURL.java:152]: Nit: space after `if`.

File: test/jdk/security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+public final class CAInterop {
+    /**
+     * Constructor for interoperability test with third party CA.
+     * @param revocationMode revocation checking mode to use
+     */
+    public CAInterop(String revocationMode){
```
Comment (seanjmullan) [inline test/jdk/security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java:28]: You could have instead added a static method to `ValidatePathWithURL`.

File: test/jdk/security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Validates provided URLs using <code>HttpsURLConnection</code> making sure they
+     * anchor to the root CA found in <code>cacerts</code> using provided alias.
+     * @param caAlias CA alis from <code>cacerts</code> file
+     * @param validCert valid test URL
+     * @param revokedCert revoked test URL
+     * @throws Exception thrown when certificate can't be validated as valid or revoked
+     */
+    public void validate(String caAlias,
```
Comment (seanjmullan) [inline test/jdk/security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java:47]: Seems like you could also add this as a static method to `ValidatePathWithURL` and then you don't need this class.

Comment (rhalade) [reply test/jdk/security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java:50]: Would prefer to leave it as a separate class for better readability.

**Review by seanjmullan — APPROVED** [review]

File: test/jdk/security/infra/java/security/cert/CertPathValidator/certification/CertignaCA.java [hunk 7]
```diff
@@ -23,70 +23,13 @@
 
 /*
  * @test
- * @bug 8245654 8314960
+ * @bug 8245654 8256895
```
Comment (rhalade) [inline test/jdk/security/infra/java/security/cert/CertPathValidator/certification/CertignaCA.java:26]: The tests for `Certigna Root CA` is moved to `CAInterop` but `Certigna` CA test will continue to use old framework as there are no test URLs.

This test is also update test for JDK-8256895 which was part of SSLCA.java test

File: test/jdk/ProblemList.txt [hunk 8]
```diff
@@ -616,6 +616,11 @@ sun/security/pkcs11/rsa/TestKeyPairGenerator.java               8295343 linux-al
 sun/security/pkcs11/rsa/TestKeyFactory.java                     8295343 linux-all
 sun/security/pkcs11/KeyStore/Basic.java                         8295343 linux-all
 
+# jdk_security_infra
+security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java#sslrootrsaca 8316381 generic-all
```
Comment (seanjmullan) [inline test/jdk/ProblemList.txt:620]: Check if these URLs work now and can be avoided putting on ProblemList.

**Review by seanjmullan — APPROVED** [review]

**Pull request closed without merging**